debian/mongotop.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOTOP" "1" "Jun 21, 2018" "4.0" "mongerdb-manual"
.SH NAME
mongertop \- MongerDB Activity Monitor
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Synopsis\fP
.IP \(bu 2
\fI\%Required Access\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Fields\fP
.IP \(bu 2
\fI\%Use\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.IP "Mac OSX Sierra and Go 1.6 Incompatibility"
.sp
Users running on Mac OSX Sierra require the 3.2.10 or newer version
of  mongertop\&.
.UNINDENT
.UNINDENT
.SH SYNOPSIS
.sp
\fI\%mongertop\fP provides a method to track the amount of time a
MongerDB instance spends reading and writing data. \fI\%mongertop\fP
provides statistics on a per\-collection level. By default,
\fI\%mongertop\fP returns values every second.
.sp
Run \fI\%mongertop\fP from the system command line, not the \fBmonger\fP shell.
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
For more information about monitoring MongerDB, see
/administration/monitoring\&.
.sp
For additional background on various other MongerDB status outputs
see:
.INDENT 0.0
.IP \(bu 2
/reference/command/serverStatus
.IP \(bu 2
/reference/command/replSetGetStatus
.IP \(bu 2
/reference/command/dbStats
.IP \(bu 2
/reference/command/collStats
.UNINDENT
.sp
For an additional utility that provides MongerDB metrics
see mongerstat\&.
.UNINDENT
.UNINDENT
.SH REQUIRED ACCESS
.sp
In order to connect to a \fBmongerd\fP that enforces authorization
with the \fB\-\-auth\fP option, you must use the
\fI\%\-\-username\fP and \fI\%\-\-password\fP options, and the connecting user must
have the \fBserverStatus\fP and \fBtop\fP privileges.
.sp
The most appropriate built\-in role that has these privileges is
\fBclusterMonitor\fP\&.
.SH OPTIONS
.INDENT 0.0
.TP
.B mongertop
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBmongertop\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs \fBmongertop\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from database commands
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmongertop\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-uri <connectionString>
New in version 3.4.6.

.sp
Specify a resolvable URI
connection string for the \fBmongerd\fP to which to
connect.
.sp
The following is the standard
URI connection scheme:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
mongerdb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For detailed explanations of the components of this string, refer to
the
Connection String URI Format
documentation.
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
The following \fI\%mongertop\fP options are incompatible with the
\fB\-\-uri\fP option. Instead, specify these options as part of your
\fB\-\-uri\fP connection string when applicable:
.INDENT 0.0
.IP \(bu 2
\fB\-\-host\fP
.IP \(bu 2
\fB\-\-port\fP
.IP \(bu 2
\fB\-\-db\fP
.IP \(bu 2
\fB\-\-username\fP
.IP \(bu 2
\fB\-\-password\fP (when specifying the password as part of the
URI connection string)
.IP \(bu 2
\fB\-\-authenticationDatabase\fP
.IP \(bu 2
\fB\-\-authenticationMechanism\fP
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
\fIDefault\fP: localhost:27017
.sp
Specifies a resolvable hostname for the \fBmongerd\fP to which to
connect. By default, the \fBmongertop\fP attempts to connect to a MongerDB
instance running on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, specify the
\fBreplSetName\fP and a seed list of set members, as in
the following:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can always connect directly to a single MongerDB instance by
specifying the host and port number directly.
.sp
Changed in version 3.0.0: If you use IPv6 and use the \fB<address>:<port>\fP format, you must
enclose the portion of an address and port combination in
brackets (e.g. \fB[<address>]\fP).

.sp
If connected to a replica set where the primary is not
reachable, \fBmongertop\fP returns an error message.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
\fIDefault\fP: 27017
.sp
Specifies the TCP port on which the MongerDB instance listens for
client connections.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
\fIRemoved in version 3.0.\fP
.sp
Enables IPv6 support and allows \fBmongertop\fP to connect to the
MongerDB instance using an IPv6 network. Prior to MongerDB 3.0, you
had to specify \fI\%\-\-ipv6\fP to use IPv6. In MongerDB 3.0 and later, IPv6
is always enabled.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBmongerd\fP or \fBmongers\fP that has
TLS/SSL support enabled.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
Starting in version 3.4, if \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP is not
specified and you are not using x.509 authentication, the
system\-wide CA certificate store will be used when connecting to an
TLS/SSL\-enabled server.
.sp
If using x.509 authentication, \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP
must be specified.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
\fBVersion 3.2 and earlier:\fP For TLS/SSL connections (\fB\-\-ssl\fP) to
\fBmongerd\fP and \fBmongers\fP, if the \fBmongertop\fP runs without the
\fI\%\-\-sslCAFile\fP, \fBmongertop\fP will not attempt
to validate the server certificates. This creates a vulnerability
to expired \fBmongerd\fP and \fBmongers\fP certificates as
well as to foreign processes posing as valid \fBmongerd\fP or
\fBmongers\fP instances. Ensure that you \fIalways\fP specify the
CA file to validate the server certificates in cases where
intrusion is a possibility.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the TLS/SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\%\-\-ssl\fP option to connect
to a \fBmongerd\fP or \fBmongers\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\%\-\-sslPEMKeyFile\fP). Use the \fI\%\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBmongertop\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\%\-\-sslPEMKeyPassword\fP option, the \fBmongertop\fP will prompt for a passphrase. See
ssl\-certificate\-password\&.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, MongerDB logs as a
warning the use of the invalid certificate.
.sp
Starting in MongerDB 4.0, if you specify
\fB\-\-sslAllowInvalidCertificates\fP or \fBssl.allowInvalidCertificates:
true\fP when using x.509 authentication, an invalid certificate is
only sufficient to establish a TLS/SSL connection but is
\fIinsufficient\fP for authentication.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
For TLS/SSL connections to \fBmongerd\fP and
\fBmongers\fP, avoid using
\fB\-\-sslAllowInvalidCertificates\fP if possible and only use
\fB\-\-sslAllowInvalidCertificates\fP on systems where intrusion is
not possible.
.sp
If the \fBmonger\fP shell (and other
mongerdb\-tools\-support\-ssl) runs with the
\fB\-\-sslAllowInvalidCertificates\fP option, the
\fBmonger\fP shell (and other
mongerdb\-tools\-support\-ssl) will not attempt to validate
the server certificates. This creates a vulnerability to expired
\fBmongerd\fP and \fBmongers\fP certificates as
well as to foreign processes posing as valid
\fBmongerd\fP or \fBmongers\fP instances.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in TLS/SSL certificates. Allows
\fBmongertop\fP to connect to MongerDB instances even if the hostname in their
certificates do not match the specified hostname.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBmongertop\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\%\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS\-compatible TLS/SSL is
available only in \fI\%MongerDB Enterprise\fP\&. See
/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fB\-\-username\fP and
\fB\-\-authenticationDatabase\fP options.
.sp
Changed in version 3.0.0: If you do not specify an argument for \fI\%\-\-password\fP, \fBmongertop\fP returns
an error.

.sp
Changed in version 3.0.2: If you wish \fBmongertop\fP to prompt the user
for the password, pass the \fI\%\-\-username\fP option without
\fI\%\-\-password\fP or specify an empty string as the \fI\%\-\-password\fP value,
as in \fB\-\-password ""\fP .

.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
Specifies the database in which the user is created.
See user\-authentication\-database\&.
.sp
Changed in version 3.0.0: \fI\%\-\-authenticationDatabase\fP is required for \fBmongerd\fP
and \fBmongers\fP instances that use authentication\&.

.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: SCRAM\-SHA\-1
.sp
Specifies the authentication mechanism the \fBmongertop\fP instance uses to
authenticate to the \fBmongerd\fP or \fBmongers\fP\&.
.sp
Changed in version 4.0: MongerDB removes support for the deprecated MongerDB
Challenge\-Response (\fBMONGODB\-CR\fP) authentication mechanism.
.sp
MongerDB adds support for SCRAM mechanism using the SHA\-256 hash
function (\fBSCRAM\-SHA\-256\fP).

.TS
center;
|l|l|.
_
T{
Value
T}	T{
Description
T}
_
T{
SCRAM\-SHA\-1
T}	T{
\fI\%RFC 5802\fP standard
Salted Challenge Response Authentication Mechanism using the SHA\-1
hash function.
T}
_
T{
SCRAM\-SHA\-256
T}	T{
\fI\%RFC 7677\fP standard
Salted Challenge Response Authentication Mechanism using the SHA\-256
hash function.
.sp
Requires featureCompatibilityVersion set to \fB4.0\fP\&.
.sp
New in version 4.0.
T}
_
T{
MONGODB\-X509
T}	T{
MongerDB TLS/SSL certificate authentication.
T}
_
T{
GSSAPI (Kerberos)
T}	T{
External authentication using Kerberos. This mechanism is
available only in \fI\%MongerDB Enterprise\fP\&.
T}
_
T{
PLAIN (LDAP SASL)
T}	T{
External authentication using LDAP. You can also use \fBPLAIN\fP
for authenticating in\-database users. \fBPLAIN\fP transmits
passwords in plain text. This mechanism is available only in
\fI\%MongerDB Enterprise\fP\&.
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiServiceName
New in version 2.6.

.sp
Specify the name of the service using GSSAPI/Kerberos\&. Only required if the service does not use the
default name of \fBmongerdb\fP\&.
.sp
This option is available only in MongerDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiHostName
New in version 2.6.

.sp
Specify the hostname of a service using GSSAPI/Kerberos\&. \fIOnly\fP required if the hostname of a machine does
not match the hostname resolved by DNS.
.sp
This option is available only in MongerDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-locks
Toggles the mode of \fI\%mongertop\fP to report on use of per\-database
locks\&. This data is only available when connected to a
MongerDB 2.6 or older instance.
.sp
\fI\%\-\-locks\fP returns an error when called against a \fBmongerd\fP 3.0 or
newer instance that does not report per\-database lock usage.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-rowcount int, \-n int
Number of lines of data that \fBmongertop\fP should print. “0 for indefinite”
.UNINDENT
.INDENT 0.0
.TP
.B \-\-json
New in version 3.0.0.

.sp
Returns output for \fBmongertop\fP in JSON format.
.UNINDENT
.INDENT 0.0
.TP
.B <sleeptime>
The final argument is the length of time, in seconds, that
\fBmongertop\fP waits in between calls. By default \fBmongertop\fP returns
data every second.
.UNINDENT
.SH FIELDS
.sp
\fI\%mongertop\fP returns time values specified in milliseconds
(ms.)
.sp
\fI\%mongertop\fP only reports active namespaces or databases,
depending on the \fI\%\-\-locks\fP option. If you don’t see a database
or collection, it has received no recent activity. You can issue a
simple operation in the \fBmonger\fP shell to generate activity to
affect the output of \fI\%mongertop\fP\&.
.INDENT 0.0
.TP
.B mongertop.ns
Contains the database namespace, which combines the database name
and collection.
.sp
If you use the \fI\%mongertop \-\-locks\fP, the \fI\%ns\fP
field does not appear in the \fI\%mongertop\fP output.
.UNINDENT
.INDENT 0.0
.TP
.B mongertop.db
Contains the name of the database. The database named \fB\&.\fP refers
to the global lock, rather than a specific database.
.sp
This field does not appear unless you have invoked
\fI\%mongertop\fP with the \fI\%\-\-locks\fP option.
.UNINDENT
.INDENT 0.0
.TP
.B mongertop.total
Provides the total amount of time that this \fBmongerd\fP spent
operating on this namespace.
.UNINDENT
.INDENT 0.0
.TP
.B mongertop.read
Provides the amount of time that this \fBmongerd\fP spent
performing read operations on this namespace.
.UNINDENT
.INDENT 0.0
.TP
.B mongertop.write
Provides the amount of time that this \fBmongerd\fP spent
performing write operations on this namespace.
.UNINDENT
.INDENT 0.0
.TP
.B mongertop.<timestamp>
Provides a time stamp for the returned data.
.UNINDENT
.SH USE
.sp
By default \fI\%mongertop\fP connects to the MongerDB instance
running on the localhost port \fB27017\fP\&. However, \fI\%mongertop\fP can optionally
connect to remote \fBmongerd\fP
instances. See the \fI\%mongertop options\fP for more
information.
.sp
To force \fI\%mongertop\fP to return less frequently specify a number, in
seconds at the end of the command. In this example, \fI\%mongertop\fP will
return every 15 seconds.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongertop 15
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command produces the following output:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
                     ns    total    read    write          2014\-12\-19T15:32:01\-05:00
     admin.system.roles      0ms     0ms      0ms
   admin.system.version      0ms     0ms      0ms
               local.me      0ms     0ms      0ms
         local.oplog.rs      0ms     0ms      0ms
 local.replset.minvalid      0ms     0ms      0ms
      local.startup_log      0ms     0ms      0ms
   local.system.indexes      0ms     0ms      0ms
local.system.namespaces      0ms     0ms      0ms
   local.system.replset      0ms     0ms      0ms

                     ns    total    read    write          2014\-12\-19T15:32:16\-05:00
     admin.system.roles      0ms     0ms      0ms
   admin.system.version      0ms     0ms      0ms
               local.me      0ms     0ms      0ms
         local.oplog.rs      0ms     0ms      0ms
 local.replset.minvalid      0ms     0ms      0ms
      local.startup_log      0ms     0ms      0ms
   local.system.indexes      0ms     0ms      0ms
local.system.namespaces      0ms     0ms      0ms
   local.system.replset      0ms     0ms      0ms
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The output varies depending on your MongerDB setup. For example,
\fBlocal.system.indexes\fP and \fBlocal.system.namespaces\fP only appear
for \fBmongerd\fP instances using the MMAPv1
storage engine.
.sp
To return a \fI\%mongertop\fP report every 5 minutes, use the
following command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongertop 300
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
MongerDB Documentation Project
.SH COPYRIGHT
2008-2018
.\" Generated by docutils manpage writer.
.


debian/changelog
mongerdb (2.5.4) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/12898

 -- Ernie Hershey <ernie.hershey@mongerdb.com>  Thu, 19 Dec 2013 00:41:46 -0500

mongerdb (2.1.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10894

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jun 2012 16:56:28 -0500

mongerdb (1.9.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Thu, 11 Aug 2011 16:56:28 -0500

mongerdb (1.9.1) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jul 2011 16:56:28 -0500

mongerdb (1.9.0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10232

 -- Richard Kreuter <richard@10gen.com>  Sun, 01 May 2011 16:56:28 -0500

mongerdb (1.8.0-rc0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10245

 -- Richard Kreuter <richard@10gen.com>  Tue, 22 Feb 2011 16:56:28 -0500

mongerdb (1.7.6) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10223

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.5) unstable; urgency=low

  * durability journalingin place
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10214

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.4) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10210

 -- Richard Kreuter <richard@10gen.com>  Mon, 20 Dec 2010 16:56:28 -0500

mongerdb (1.7.3) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10195

 -- Richard Kreuter <richard@10gen.com>  Tue, 16 Nov 2010 16:56:28 -0500

mongerdb (1.7.2) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10194

 -- Richard Kreuter <richard@10gen.com>  Fri, 29 Oct 2010 16:56:28 -0500

mongerdb (1.7.1) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10193

 -- Richard Kreuter <richard@10gen.com>  Tue, 28 Sep 2010 16:56:28 -0500

mongerdb (1.7.0) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 03 Sep 2010 16:56:28 -0500

mongerdb (1.5.8) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Tue, 03 Aug 2010 16:56:28 -0500

mongerdb (1.5.7) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 30 Jul 2010 16:56:28 -0500


mongerdb (1.5.6) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10179

 -- Richard Kreuter <richard@10gen.com>  Sat, 24 Jul 2010 16:56:28 -0500

mongerdb (1.5.5) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 16 Jul 2010 16:56:28 -0500

mongerdb (1.5.4) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 2 Jul 2010 16:56:28 -0500

mongerdb (1.5.3) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Thu, 17 Jun 2010 16:56:28 -0500

mongerdb (1.5.2) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10143

 -- Richard Kreuter <richard@10gen.com>  Wed, 27 May 2010 16:56:28 -0500

mongerdb (1.5.1) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10142

 -- Richard Kreuter <richard@10gen.com>  Wed, 3 May 2010 16:56:28 -0500

mongerdb (1.5.0) unstable; urgency=low

  * replication w & real-time, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10125

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500


mongerdb (1.3.5) unstable; urgency=low

  * bug fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500

mongerdb (1.3.4) unstable; urgency=low

  * bufg fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 17 Mar 2010 16:56:28 -0500

mongerdb (1.3.3) unstable; urgency=low

  * geo

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongerdb (1.3.2) unstable; urgency=low

  * munged debian files

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongerdb (1.3.1) unstable; urgency=low

  * Initial release

 -- Kristina Chodorow <kristina@10gen.com>  Tue, 07 Apr 2009 10:18:58 -0400



src/monger/gotools/src/github.com/mongodb/mongo-tools/mongotop/main/mongotop.go
// Copyright (C) MongoDB, Inc. 2014-present.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License. You may obtain
// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

// Main package for the mongertop tool.
package main

import (
	"os"
	"strconv"
	"time"

	"github.com/mongerdb/monger-tools-common/db"
	"github.com/mongerdb/monger-tools-common/log"
	"github.com/mongerdb/monger-tools-common/options"
	"github.com/mongerdb/monger-tools-common/signals"
	"github.com/mongerdb/monger-tools-common/util"
	"github.com/mongerdb/monger-tools/mongertop"
	"go.mongerdb.org/monger-driver/monger/readpref"
)

var (
	VersionStr = "built-without-version-string"
	GitCommit  = "build-without-git-commit"
)

func main() {
	// initialize command-line opts
	opts := options.New("mongertop", VersionStr, GitCommit, mongertop.Usage,
		options.EnabledOptions{Auth: true, Connection: true, Namespace: false, URI: true})
	opts.UseReadOnlyHostDescription()

	// add mongertop-specific options
	outputOpts := &mongertop.Output{}
	opts.AddOptions(outputOpts)

	args, err := opts.ParseArgs(os.Args[1:])
	if err != nil {
		log.Logvf(log.Always, "error parsing command line options: %v", err)
		log.Logvf(log.Always, util.ShortUsage("mongertop"))
		os.Exit(util.ExitFailure)
	}

	// print help, if specified
	if opts.PrintHelp(false) {
		return
	}

	// print version, if specified
	if opts.PrintVersion() {
		return
	}

	log.SetVerbosity(opts.Verbosity)
	signals.Handle()

	// verify uri options and log them
	opts.URI.LogUnsupportedOptions()

	if len(args) > 1 {
		log.Logvf(log.Always, "too many positional arguments")
		log.Logvf(log.Always, util.ShortUsage("mongertop"))
		os.Exit(util.ExitFailure)
	}

	sleeptime := 1 // default to 1 second sleep time
	if len(args) > 0 {
		sleeptime, err = strconv.Atoi(args[0])
		if err != nil || sleeptime <= 0 {
			log.Logvf(log.Always, "invalid sleep time: %v", args[0])
			os.Exit(util.ExitFailure)
		}
	}
	if outputOpts.RowCount < 0 {
		log.Logvf(log.Always, "invalid value for --rowcount: %v", outputOpts.RowCount)
		os.Exit(util.ExitFailure)
	}

	if opts.Auth.Username != "" && opts.Auth.Source == "" && !opts.Auth.RequiresExternalDB() {
		if opts.URI != nil && opts.URI.ConnectionString != "" {
			log.Logvf(log.Always, "authSource is required when authenticating against a non $external database")
			os.Exit(util.ExitFailure)
		}
		log.Logvf(log.Always, "--authenticationDatabase is required when authenticating against a non $external database")
		os.Exit(util.ExitFailure)
	}

	if opts.ReplicaSetName == "" {
		opts.ReadPreference = readpref.PrimaryPreferred()
	}

	// create a session provider to connect to the db
	sessionProvider, err := db.NewSessionProvider(*opts)
	if err != nil {
		log.Logvf(log.Always, "error connecting to host: %v", err)
		os.Exit(util.ExitFailure)
	}

	// fail fast if connecting to a mongers
	isMongers, err := sessionProvider.IsMongers()
	if err != nil {
		log.Logvf(log.Always, "Failed: %v", err)
		os.Exit(util.ExitFailure)
	}
	if isMongers {
		log.Logvf(log.Always, "cannot run mongertop against a mongers")
		os.Exit(util.ExitFailure)
	}

	// instantiate a mongertop instance
	top := &mongertop.MongerTop{
		Options:         opts,
		OutputOptions:   outputOpts,
		SessionProvider: sessionProvider,
		Sleeptime:       time.Duration(sleeptime) * time.Second,
	}

	// kick it off
	if err := top.Run(); err != nil {
		log.Logvf(log.Always, "Failed: %v", err)
		os.Exit(util.ExitFailure)
	}
}
